src/picongpu/include/particles/Particles.kernel

```diff
@@ -255,93 +255,186 @@ struct KernelManipulateAllParticles
+        DataSpace< simDim > const superCellOffset = block * SuperCellSize::toRT();
+
+         using ParticleDomCfg = IdxConfig<
+            frameSize,
+            numWorkers
+        >;
+
+        PMACC_SMEM(
+            mustShift,
+            int
+        );
+
+        /* unique for each worker */
+        FramePtr frame;
+        lcellId_t particlesInSuperCell;
+
+        ForEachIdx<
+            IdxConfig<
+                1,
+                numWorkers
+            >
+        > onlyMaster{ workerIdx };
+
+        onlyMaster(
+            [&](
+                uint32_t const,
+                uint32_t const
+            )
+            {
+                mustShift = 0;
+            }
+        );
+
+        frame = pb.getLastFrame( block );
+        particlesInSuperCell = pb.getSuperCell( block ).getSizeLastFrame( );
+
+        auto cachedB = CachedBox::create <
+            0,
+            typename T_BBox::ValueType
+        >( T_DataDomain() );
+        auto cachedE = CachedBox::create <
+            1,
+            typename T_EBox::ValueType
+        >( T_DataDomain( ) );
+
+        __syncthreads();
+
+        //end kernel if we have no frames
+        if( !frame.isValid( ) )
+           return;
+
+        nvidia::functors::Assign assign;
+        ThreadCollective<
+            T_DataDomain,
+            numWorkers
+        > collective{ workerIdx };
+
+        auto fieldBBlock = fieldB.shift( superCellOffset );
+        collective(
+            assign,
+            cachedB,
+            fieldBBlock
+        );
+
+        auto fieldEBlock = fieldE.shift( superCellOffset );
+        collective(
+            assign,
+            cachedE,
+            fieldEBlock
+        );
+
+        __syncthreads();
+
+        // move over frames and call frame solver
+        while( frame.isValid( ) )
+        {
+            // loop over all particles in the frame
+            ForEachIdx< ParticleDomCfg >{ workerIdx }
+            (
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    if( linearIdx < particlesInSuperCell )
+                    {
+                        particleFunctor(
+                            *frame,
+                            linearIdx,
+                            cachedB,
+                            cachedE,
+                            mustShift
+                        );
+                    }
+                }
+            );
+            // independent for each worker
+            frame = pb.getPreviousFrame( frame );
+            particlesInSuperCell = frameSize;
+        }
+
+        __syncthreads();
+
+        onlyMaster(
+            [&](
+                uint32_t const,
+                uint32_t const
+            )
+            {
+                /* set in SuperCell the mustShift flag which is a optimization
+                 * for shift particles and fillGaps
```

Comment:
maybe add what this optimizes at the place where `mustShift` is defined in the supercell